Kernel/VirtIO/VirtIO.h

```diff
@@ -0,0 +1,251 @@
+#define COMMON_CFG_QUEUE_SIZE 0x18
+#define COMMON_CFG_QUEUE_MSIX_VECTOR 0x1a
+#define COMMON_CFG_QUEUE_ENABLE 0x1c
+#define COMMON_CFG_QUEUE_NOTIFY_OFF 0x1e
+#define COMMON_CFG_QUEUE_DESC 0x20
+#define COMMON_CFG_QUEUE_DRIVER 0x28
+#define COMMON_CFG_QUEUE_DEVICE 0x30
+
+#define QUEUE_INTERRUPT 0x1
+#define DEVICE_CONFIG_INTERRUPT 0x2
+
+class VirtIO {
+public:
+    static void detect();
+};
+
+class VirtIODevice : public PCI::Device {
+public:
+    VirtIODevice(PCI::Address, const char*);
+    virtual ~VirtIODevice() override;
+
+protected:
+    const char* const m_class_name;
+
+    struct MappedMMIO {
+        OwnPtr<Region> base;
+        size_t size { 0 };
+
+        template<typename T>
+        T read(u32 offset) const
+        {
+            if (!base)
+                return 0;
+            ASSERT(size >= sizeof(T));
+            ASSERT(offset + sizeof(T) <= size);
+            return *(volatile T*)(base->vaddr().offset(offset).get());
+        }
+
+        template<typename T>
+        void write(u32 offset, T value)
+        {
+            if (!base)
+                return;
+            ASSERT(size >= sizeof(T));
+            ASSERT(offset + sizeof(T) <= size);
+            *(volatile T*)(base->vaddr().offset(offset).get()) = value;
+        }
+    };
+
+    struct Configuration {
+        u8 cfg_type;
+        u8 bar;
+        u32 offset;
+        u32 length;
+    };
+
+    const Configuration* get_config(u8 cfg_type, u32 index = 0) const
+    {
+        for (const auto& cfg : m_configs) {
+            if (cfg.cfg_type != cfg_type)
+                continue;
+            if (index > 0) {
+                index--;
+                continue;
+            }
+            return &cfg;
+        }
+        return nullptr;
+    }
+    const Configuration* get_common_config(u32 index = 0) const;
+    const Configuration* get_device_config(u32 index = 0) const;
+
+    template<typename F>
+    void read_config_atomic(F f)
+    {
+        if (m_common_cfg) {
+            u8 generation_before, generation_after;
+            do {
+                generation_before = config_read8(m_common_cfg, 0x15);
+                f();
+                generation_after = config_read8(m_common_cfg, 0x15);
+            } while (generation_before != generation_after);
+        } else {
+            f();
+        }
+    }
+
+    u8 config_read8(const Configuration*, u32);
+    u16 config_read16(const Configuration*, u32);
+    u32 config_read32(const Configuration*, u32);
+    void config_write8(const Configuration*, u32, u8);
+    void config_write16(const Configuration*, u32, u16);
+    void config_write32(const Configuration*, u32, u32);
+    void config_write64(const Configuration*, u32, u64);
+
+    auto mapping_for_bar(u8) -> MappedMMIO&;
+
+    u8 read_status_bits();
+    void clear_status_bit(u8);
+    void set_status_bit(u8);
+    u64 get_device_features();
+    bool finish_init();
+
+    VirtIOQueue* get_queue(u16 queue_index)
```

Comment:
Either use the `OwnPtr` container, or return a reference and put in the first line of the method `ASSERT` statement.